WAN federation docs (#20)
boruszak/consul-ia-experiments: website/content/docs/east-west/wan-federation/vms.mdx

```shell-session
$ curl http://localhost:8500/v1/catalog/nodes?dc=dc2
```

This will return all the known nodes in the `dc2` datacenter.

<CodeBlockConfig hideClipboard highlight="6">

```json
[
  {
    "ID": "a3727eaa-da9d-8dda-07ff-97b7b451415d",
    "Node": "consul-server-1",
    "Address": "172.18.0.7",
    "Datacenter": "dc2",
    "TaggedAddresses": {
      "lan": "172.18.0.7",
      "lan_ipv4": "172.18.0.7",
      "wan": "172.18.0.7",
      "wan_ipv4": "172.18.0.7"
    },
    "Meta": {
      "consul-network-segment": "",
      "consul-version": "1.20.2"
    },
    "CreateIndex": 13,
    "ModifyIndex": 15
  }
]
```

</CodeBlockConfig>

</Tab>
<Tab heading="UI" group="ui">

Federated datacenters show a dropdown on the UI that allows to select the datacenter for which to show services and nodes.

![Consul UI showing service page with WAN dropdown](/img/east-west/wan-federation/consul-ui-services-multi_dc_selector-dark.png#dark-theme-only)
![Consul UI showing service page with WAN dropdown](/img/east-west/wan-federation/consul-ui-services-multi_dc_selector.png#light-theme-only)

</Tab>
</Tabs>

## Troubleshooting WAN federation

> Configuring `serf_wan`, `advertise_addr_wan` and `translate_wan_addrs` can lead to a situation where `consul members -wan` lists remote nodes but RPC 
> operations fail with one of the following errors:

Reviewer: ```suggestion
Configuring `serf_wan`, `advertise_addr_wan` and `translate_wan_addrs` can lead to a situation where `consul members -wan` lists remote nodes but RPC operations fail with one of the following errors:
```

- `No path to datacenter`
- `rpc error getting client: failed to get conn: dial tcp <LOCAL_ADDR>:0-><REMOTE_ADDR>:<REMOTE_RPC_PORT>: i/o timeout`

The most likely cause of these errors is that `bind_addr` is set to a private address preventing the RPC server from accepting connections across the WAN. Setting `bind_addr` to a public address (or one that can be routed across the WAN) will resolve this issue. Be aware that exposing the RPC server on a public port should only be done **after** firewall rules have been established.

The [`translate_wan_addrs`](/consul/docs/reference/agent#translate_wan_addrs) configuration provides a basic address rewriting capability.

## Data replication

> In general, data is not replicated between different Consul datacenters. When a request is made for a resource in another datacenter, the local Consul servers forward an RPC request to the remote Consul servers for that resource and return the results. If the remote datacenter is not available, then those resources will also not be available, but that won't otherwise affect the local datacenter. 
> There are some special situations where a limited subset of data can be replicated, such as with Consul's built-in ACL replication capability, or
> external tools like [consul-replicate](https://github.com/hashicorp/consul-replicate/).

Reviewer: ```suggestion
In general, data is not replicated between different Consul datacenters. When a request is made for a resource in another datacenter, the local Consul servers forward an RPC request to the remote Consul servers for that resource and return the results. If the remote datacenter is not available, then those resources will also not be available. However, this will not otherwise affect the local datacenter. 

There are some special situations where a limited subset of data can be replicated, such as with Consul's built-in ACL replication capability, or external tools like [consul-replicate](https://github.com/hashicorp/consul-replicate/).
```
